File: src/backend/dotNet/dotNet/Models/ANNSettings.cs


namespace dotNet.Models {
    public class ANNSettings {

        // General info
        public ProblemType ANNType;

        public ANNSettings(
            ProblemType aNNType, 
            float learningRate, 
            int batchSize, int 
            numberOfEpochs, 
            int inputSize, 
            int outputSize, 
            int[]? hiddenLayers, 
            ActivationFunction[]? activationFunctions,
            RegularizationMethod regularization, 
            LossFunction lossFunction, 
            Optimizer optimizer) {

            ANNType = aNNType;
            LearningRate = learningRate;
            BatchSize = batchSize;
            NumberOfEpochs = numberOfEpochs;
            InputSize = inputSize;
            OutputSize = outputSize;
            HiddenLayers = hiddenLayers;
            ActivationFunctions = activationFunctions;
            Regularization = regularization;
            LossFunction = lossFunction;
            Optimizer = optimizer;
        }

        public float LearningRate { get; set;}
        public int BatchSize { get; set; }
        public int NumberOfEpochs { get; set; }
        // IO
        public int InputSize { get; set; }
        public int OutputSize { get; set; }
        // Layers
        public int[]? HiddenLayers { get; set; }
        public ActivationFunction[]? ActivationFunctions { get; set; }
        // Other
        public RegularizationMethod Regularization { get; set; }
        public LossFunction LossFunction { get; set; }
        public Optimizer Optimizer { get; set; }

        // To JSON string
        public override string ToString() {
            return Newtonsoft.Json.JsonConvert.SerializeObject(this);
        }
    }

    public enum ProblemType {
        Regression,
        Classification
    }

    public enum ActivationFunction {
        ReLU,
        LeakyReLU,
        Sigmoid,
        Tanh,
        Linear
    }

    public enum RegularizationMethod {
        L1,
        L2
    }

    public enum LossFunction {
        L1Loss,
        MSELoss,
        CrossEntropyLoss
    }

    public enum Optimizer {
        SGD,
        Adagrad,
        Adadelta,
        Adam
    }
}

File: src/ml/ANNSettings.py

import json


class ANNSettings:
    
    
    def __init__(
        self,
        problemType         = 0, 
        learningRate        = 0.0, 
        batchSize           = 0, 
        numberOfEpochs      = 0, 
        inputSize           = 0, 
        outputSize          = 0, 
        hiddenLayers        = None, 
        activationFunctions = None,
        regularization      = 0,
        lossFunction        = 0,
        optimizer           = 0
        ) -> None:
        
        self.problemType         = problemType
        self.learningRate        = learningRate
        self.batchSize           = batchSize
        self.numberOfEpochs      = numberOfEpochs
        self.inputSize           = inputSize
        self.outputSize          = outputSize
        self.hiddenLayers        = hiddenLayers
        self.activationFunctions = activationFunctions
        self.regularization      = regularization
        self.lossFunction        = lossFunction
        self.optimizer           = optimizer
    
    def load(data) -> None:
        jsonObj = json.loads(data)
        return ANNSettings(
            problemType         = jsonObj["ANNType"],
            learningRate        = jsonObj["LearningRate"],
            batchSize           = jsonObj["BatchSize"],
            numberOfEpochs      = jsonObj["NumberOfEpochs"],
            inputSize           = jsonObj["InputSize"],
            outputSize          = jsonObj["OutputSize"],
            hiddenLayers        = jsonObj["HiddenLayers"],
            activationFunctions = jsonObj["ActivationFunctions"],
            regularization      = jsonObj["Regularization"],
            lossFunction        = jsonObj["LossFunction"],
            optimizer           = jsonObj["Optimizer"]
        )
    
    
    